### ISSUE TITLE
List and Array extras shouldn't be in "Extensions"

### ISSUE THREAD

**Issue by mlms13** [reazen/relude#176]
We seem to be using `Relude_Extensions_*` to mean two different things. For the most part it's used to represent "functions that come for free from typeclass implementations", but in the case of `Relude_Extensions_List.re` and `Relude_Extensions_Array.re`, it's used to mean "extra List and Array functions based on the inner type". This means there are lots of functions that can be accessed a couple different ways:

```reason
Relude.Array.Float.contains;
Relude.Extensions.Array.Float.contains;
```

I don't think we've ever highlighted that these things are in `Relude.Extensions`, and I think it's more confusing than helpful to lump them in with typeclass-related functions. I'm inclined to remove the `Extensions` versions and move the file into the appropriate list/array folders, but I thought I'd mention it here first in case there are any objections.

**Comment by andywhite37** [comment]
Yeah, I think that makes sense.  I think at the time, I was trying to organize and centralize all the stuff we were `including` into other modules, but you're right, the `List`/`Array` stuff is different - it's more like specializations that extensions.  Maybe `Specializations` even makes sense as a name for those things?

On a separate note, I'm still not 100% happy with the naming conventions for the extensions.  Maybe it would make sense to un-nest some of those (mostly `Foldable`), and make them more specialized, like `Relude_FoldableExtensions.Make` and `Relude_FoldableInfix.Make` and pull out the functors that are defined inside of those into separate things, so we don't have all the weird naming conflict issues we have now, when you include something in and in implicitly brings in more module functors.

Given our bundling woes and just the shortcomings of OCaml's modules and impurity, I'm starting to wonder if auto-including stuff into modules is something we need to re-think.  Maybe it would be better to put that burden onto the users of the library to assemble the pieces they need, rather than trying to give it all to them pre-baked.

I haven't put a ton of thought into all this, and if we are going to undertake a major re-org, I'll be sure to put out some more concrete ideas first.

**Comment by andywhite37** [comment]
I went ahead and moved these into the `array` and `list` folders, and I named them `Relude_List_Specializations` and same for `Array` - I wanted to undertake the task to remove the .rei files for `Array` and `List`, so it made sense to do this while I was at it.  I chose that name for lack of a better idea, but we can rename them something different if anyone has a better idea.  I thought it was useful to differentiate the name `Extensions` (which are what we are calling module functors that give you extra functions when you have an instance of a typeclass), from these things, which are functions we can give you for specializations of `List` or `Array`, like `List.String.join`, etc.

Released in 0.33.0
